### PR TITLE
Fixed the server and added some base features for demos

### PR DESCRIPTION
**Pull request by jmwright**
With the exception of allowing download of individual components (which might end up being a #wontfix anyway), the server is fixed and has a couple of new features.

1. The component drop downs are now populated based on the `devices.json` file, and any components that are missing necessary metadata ("HeightUnits") are filtered out so that you cannot generate a broken config.
2. There is a 3D preview that loads the generated glTF (*.glb) file for the assembly. This 3D preview should dynamically update whenever a new build is finished generating (at the same time the zip file should download).

This is a contrived screenshot. The preview actually works, and has full 3D mouse controls.
Screenshot from 2024-05-21 14-38-47

This server is still a mess in some ways. The `generate` call blocks, and special care will be needed to make that work async without breaking the server framework (it is weird about spinning off threads from within it's own threads). This means that it is not ready for multiple simultaneous users. However, it should work fine for a local demo on a laptop.

Caching is still in place, but works differently than before. Both the build zip file and the glTF file are both cached.

@julianstirling If you get a chance, you could see if you can get this running on your local machine. The `requirements.txt` in the `server` directory has been updated, and is what I am running on.

### CODE REVIEW

**Comment by julianstirling** [comment]
Also.. the preview is upside down, but the controls are gimbal locked so you can't spin it the right way up.

**Comment by jmwright** [comment]
> Is it possible to add the assembly-docs folder to the cache? And then make a simple server route that serves from this directory?

That should be possible. I'll give it a try. 

> Also.. the preview is upside down, but the controls are gimbal locked so you can't spin it the right way up.

Yeah, I haven't figured out the model-viewer options to fix the gimbal lock yet. I can at least set the view to be right side up by default with the view direction settings.

**Comment by jmwright** [comment]
@julianstirling I fixed the issue with the preview being upside down. Hosting the resulting `assembly-docs` directory will take a little longer.

Screenshot from 2024-05-22 06-57-37

**Comment by jmwright** [comment]
Something else that should not be too hard, is to start with no drop downs. There would be an add button for each component type, something like `Add Access Point`, `Add Router`, `Add Server` and `Add Switch`. Clicking an `Add *` button would add a corresponding drop down. Then the rack can be built in any order, with any mix of components.

**Comment by jmwright** [comment]
@julianstirling Your suggestions have been implemented.

1. Fix the view being upside down.
2. Add a button to view the generated assembly documentation.

In  addition, I have renamed `test_page` to `configurator` because I think we've gone past the `test_page` stage.

Screenshot from 2024-05-22 10-02-16

I think I'll go ahead and take a shot at making the drop downs configurable.

**Comment by jmwright** [comment]
Ok, there are improvements that could be made, but I am going to stop and wait for feedback. The UI is dynamic now, allowing for arbitrarily complex racks to be configured. Setting a drop down to `None` will cause it to be removed/ignored. There is no reordering of rack components, so you have to remove components and re-add them if you make a mistake in ordering.

One question I have is whether or not the rack's CAD model is being built in the right order relative to the order of the drop downs in the UI. I'm not sure if it should be bottom-up or top-down relative to how the controls are arranged in the UI.

Screenshot from 2024-05-22 11-30-11

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:22]: At some point we probably want to move the generate functionality into the nimble_orchestration module so we don't have do do weird path things. `generate.py` can then basically become a 2-3 line script.

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:40]: I like the term unique_name. I think it is clear and unambiguous.

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:44]: `unique_name = "".join(config)`
could replace lines 44-48
`unique_name = "!".join(config)`
could be used to use ! as a delimiter. If wanted.

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:98]: Maybe remove commented out code or say why it is commented out?

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:110]: Again maybe remove commented out code.

File: server/nimble_server.py
Comment (julianstirling) [inline server/nimble_server.py:151]: Best practice is to explicitly specify encoding with `open('../devices.json', 'r', encoding='utf-8')`

**Review by julianstirling — APPROVED** [review]
A few little comments that are optional.

How do I actually suggest code changes rather than just adding comments?

**Comment by jmwright** [comment]
@julianstirling If you want to completely clear the cache during testing (start from scratch), delete the contents of the `_cache_` and `server/static/builds` directories. I copy the cached `assembly-docs` directories to the static directory for now rather than creating dynamic static directories via the web framework.

The cache is keyed on the existence of the build zip file that gets sent to the client, so a dirty cache clear can be completed by deleting the zip file, which should trigger a new build. All the other directories for that specific configuration should be re-created. If the order of the components in the config is changed, it should be treated as a completely different build.

**Comment by jmwright** [comment]
> How do I actually suggest code changes rather than just adding comments?

Click the commit diff button shown below and then edit the contents inside the `suggestion` block.

Screenshot from 2024-05-23 07-48-44

File: server/nimble_server.py
Comment (jmwright) [reply server/nimble_server.py:44]: The delimiter needs to be acceptable in filenames, as this helps make the filesystem caching mechanism work.

Comment (julianstirling) [reply server/nimble_server.py:44]: `!` should be safe in windows and unix which is why i picked it over more conventional delimiters

File: server/nimble_server.py
Comment (jmwright) [reply server/nimble_server.py:22]: Issue created here: https://github.com/Wakoma/nimble/issues/72

File: server/nimble_server.py
Comment (jmwright) [reply server/nimble_server.py:98]: This commented code is a placeholder for how the dynamic static serving of the cached assembly-docs could/should? work. I'd like to leave it in for one more revision.

Comment (julianstirling) [reply server/nimble_server.py:98]: Maybe we just add a comment to say that so it is clear?

File: server/nimble_server.py
Comment (jmwright) [reply server/nimble_server.py:40]: Do we want to make an issue to move to this nomenclature? Is it replacing, the "hash" terminology?

Comment (julianstirling) [reply server/nimble_server.py:40]: The references to configuration hashes have already been removed from orchestration, as we stopped hashing there. It may be easier to change the four instances in `index.js` as part of this PR?

**Comment by julianstirling** [comment]
Perfect. Merge is approved so feel free to merge when you are ready.

**Pull request merged**
